src/java/org/apache/cassandra/index/sai/disk/format/IndexDescriptor.java

```diff
@@ -0,0 +1,530 @@
+        }
+    }
+
+    @Override
+    public int hashCode()
+    {
+        return Objects.hashCode(descriptor, version);
+    }
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        IndexDescriptor other = (IndexDescriptor)o;
+        return Objects.equal(descriptor, other.descriptor) &&
+               Objects.equal(version, other.version);
+    }
+
+    @Override
+    public String toString()
+    {
+        return descriptor.toString() + "-SAI";
+    }
+
+    public String logMessage(String message)
+    {
+        // Index names are unique only within a keyspace.
+        return String.format("[%s.%s.*] %s",
+                             descriptor.ksname,
+                             descriptor.cfname,
+                             message);
+    }
+
+    private void registerPerSSTableComponents()
+    {
+        version.onDiskFormat()
+               .perSSTableComponents()
+               .stream()
+               .filter(c -> !perSSTableComponents.contains(c) && fileFor(c).exists())
+               .forEach(perSSTableComponents::add);
+    }
+
+    private void registerPerIndexComponents(IndexContext indexContext)
+    {
+        Set<IndexComponent> indexComponents = perIndexComponents.computeIfAbsent(indexContext.getIndexName(), k -> Sets.newHashSet());
+        version.onDiskFormat()
+               .perIndexComponents(indexContext)
+               .stream()
+               .filter(c -> !indexComponents.contains(c) && fileFor(c, indexContext).exists())
+               .forEach(indexComponents::add);
+    }
+
+    private int numberOfComponents(IndexContext indexContext)
+    {
+        return perIndexComponents.containsKey(indexContext.getIndexName()) ? perIndexComponents.get(indexContext.getIndexName()).size() : 0;
+    }
+
+    private File tmpFileFor(IndexComponent component)
+    {
+        return onDiskPerSSTableTemporaryFileMap.computeIfAbsent(component,
+                                                                c -> createFile(component, null, true));
+    }
+
+    private File tmpFileFor(IndexComponent component, IndexContext indexContext)
+    {
+        return onDiskPerIndexTemporaryFileMap.computeIfAbsent(Pair.create(component, indexContext.getIndexName()),
+                                                      c -> createFile(component, indexContext, true));
+    }
+
+    private File createFile(IndexComponent component, IndexContext indexContext, boolean temporary)
+    {
+        Component customComponent = new Component(Component.Type.CUSTOM, componentName(component, indexContext));
+        return temporary ? new File(descriptor.tmpFilenameFor(customComponent)) : new File(descriptor.filenameFor(customComponent));
+    }
+
+    private void deleteComponent(File file)
+    {
+        logger.debug("Deleting storage attached index component file {}", file);
```

Comment:
super-nit: `...storage-attached index...`?

Reply:
I was about to say that we use this elsewhere in comments and log messages but there seems to be a majority split in favour of `storage-attached` so I will change any occurrences to that. This should also be using `logMessage` to properly format the message.